### PR TITLE
change: version format

### PR DESCRIPTION
**Pull request by MrIvanPlays**
Related to #105 

This updates the version inside plugin.yml to `major.minor.patch-git.commit.id.abbrev-RELEASE/SNAPSHOT`. 

Also added an updateVersion script so it is easier to change versions :) .

### CODE REVIEW

File: bukkit/pom.xml
Comment (MrNemo64) [inline bukkit/pom.xml:86]: Would it be worth to have properties for each api module and use the properties on dependencies instead of putting just the version?

Comment (MrIvanPlays) [reply bukkit/pom.xml:86]: couldn't understand your question ; do you mean having a property which holds the maven api version so it can be used in dependencies?

**Review by MrNemo64 — APPROVED** [review]

**Review by lokka30 — APPROVED** [review]
Thank you very much Ivan! Would like to hear from Jikoo if possible, before merging. :)

**Comment by MrIvanPlays** [comment]
I am going to do the changes for the update checker too right now so if jikoo approves don't merge yet.

File: core/src/main/java/me/lokka30/treasury/plugin/core/utils/PluginVersion.java
Comment (lokka30) [inline core/src/main/java/me/lokka30/treasury/plugin/core/utils/PluginVersion.java:31]: Perhaps since users will see this, I guess it's better to say `Your current Treasury version is not formatted correctly! Please inform a Treasury maintainer as soon as possible.`

**Comment by Jikoo** [comment]
Please don't wait on my approval - I don't have a PC for a week.

File: core/src/main/java/me/lokka30/treasury/plugin/core/utils/PluginVersion.java
Comment (lokka30) [reply core/src/main/java/me/lokka30/treasury/plugin/core/utils/PluginVersion.java:31]: (This request also applies to the other usages of `Illegal Treasury version found` throughout this PR)

File: core/src/main/java/me/lokka30/treasury/plugin/core/utils/UpdateChecker.java
Comment (MrIvanPlays) [reply core/src/main/java/me/lokka30/treasury/plugin/core/utils/UpdateChecker.java:29]: there is already a todo added

File: core/src/main/java/me/lokka30/treasury/plugin/core/utils/UpdateChecker.java
Comment (MrIvanPlays) [reply core/src/main/java/me/lokka30/treasury/plugin/core/utils/UpdateChecker.java:48]: it doesn't matter at all where this check would be done in the method or before it getting called. I put it here just in case somehow there is a method to support the SRP (Single Responsibility Principle) in any of the platforms we plan to support.

Comment (lokka30) [reply core/src/main/java/me/lokka30/treasury/plugin/core/utils/UpdateChecker.java:48]: I disagree - I think it makes the code cleaner. Otherwise anywhere the update check method is called, we will also have to include that check.

File: core/src/main/java/me/lokka30/treasury/plugin/core/utils/UpdateChecker.java
Comment (lokka30) [reply core/src/main/java/me/lokka30/treasury/plugin/core/utils/UpdateChecker.java:29]: I apologize, either I missed it or the diff tool didn't show it.

File: core/src/main/java/me/lokka30/treasury/plugin/core/utils/UpdateChecker.java
Comment (MrNemo64) [inline core/src/main/java/me/lokka30/treasury/plugin/core/utils/UpdateChecker.java:87]: How does this get our commit date?

Comment (MrIvanPlays) [reply core/src/main/java/me/lokka30/treasury/plugin/core/utils/UpdateChecker.java:87]: I've configured the maven jar plugin to create a manifest file containing an entry "Implementation-Version" which is populated via maven with the commit date.

**Comment by MrIvanPlays** [comment]
If there isn't anything else I'm merging this.

**Review by lokka30 — APPROVED** [review]
LGTM. Once @MrNemo64 is satisfied with it, feel free to merge whenever you are comfortable to :)

File: core/src/main/java/me/lokka30/treasury/plugin/core/utils/UpdateChecker.java
Comment (MrNemo64) [inline core/src/main/java/me/lokka30/treasury/plugin/core/utils/UpdateChecker.java:62]: This message will be seem by server owners, they may not even know what IO is. Maybe a more user friendly message is "Exception while getting plugin data from spigot"

Comment (lokka30) [reply core/src/main/java/me/lokka30/treasury/plugin/core/utils/UpdateChecker.java:62]: Suggestion: `Unable to retrieve the latest version data for the Treasury resource on SpigotMC.org. Is your internet connection experiencing any trouble?`

File: core/src/main/java/me/lokka30/treasury/plugin/core/utils/UpdateChecker.java
Comment (MrNemo64) [reply core/src/main/java/me/lokka30/treasury/plugin/core/utils/UpdateChecker.java:90]: I see

File: core/src/main/java/me/lokka30/treasury/plugin/core/utils/UpdateChecker.java
Comment (lokka30) [reply core/src/main/java/me/lokka30/treasury/plugin/core/utils/UpdateChecker.java:48]: @MrNemo64 is this conversation resolved?

**Review by lokka30 — APPROVED** [review]

File: core/src/main/java/me/lokka30/treasury/plugin/core/utils/UpdateChecker.java
Comment (MrNemo64) [inline core/src/main/java/me/lokka30/treasury/plugin/core/utils/UpdateChecker.java:137]: This line is the same as the line 125, why have it 2 times? Is it for redability or something?

File: core/src/main/java/me/lokka30/treasury/plugin/core/utils/UpdateChecker.java
Comment (MrNemo64) [inline core/src/main/java/me/lokka30/treasury/plugin/core/utils/UpdateChecker.java:161]: I can understand the warn message but why is "USE WITH COUTION" at the error level? It makes more sense on the warn level

File: core/src/main/java/me/lokka30/treasury/plugin/core/utils/PluginVersion.java
Comment (MrNemo64) [inline core/src/main/java/me/lokka30/treasury/plugin/core/utils/PluginVersion.java:41]: Instead of all these checks and throws we could use regex to check if the version format is valid

Comment (MrIvanPlays) [reply core/src/main/java/me/lokka30/treasury/plugin/core/utils/PluginVersion.java:41]: Regex is something I don't want to get involved with. If you give me the pattern and the code to handle it, sure I'd use it. But I am not implementing such thing at all.

File: core/src/main/java/me/lokka30/treasury/plugin/core/utils/UpdateChecker.java
Comment (MrNemo64) [reply core/src/main/java/me/lokka30/treasury/plugin/core/utils/UpdateChecker.java:161]: I've seen that you use more times that technique. Since color coding seems to be supported by the logger (in the line 246 of this file color coding is ussed) couldn't you color code the message?

File: core/src/main/java/me/lokka30/treasury/plugin/core/utils/UpdateChecker.java
Comment (MrNemo64) [reply core/src/main/java/me/lokka30/treasury/plugin/core/utils/UpdateChecker.java:48]: Yes

File: core/src/main/java/me/lokka30/treasury/plugin/core/utils/UpdateChecker.java
Comment (MrNemo64) [reply core/src/main/java/me/lokka30/treasury/plugin/core/utils/UpdateChecker.java:58]: Yes

File: core/src/main/java/me/lokka30/treasury/plugin/core/utils/UpdateChecker.java
Comment (MrIvanPlays) [reply core/src/main/java/me/lokka30/treasury/plugin/core/utils/UpdateChecker.java:161]: Why do you care so much about what log level is used? Isn't it better to have the entire line colored in red rather than a part of it?

File: core/src/main/java/me/lokka30/treasury/plugin/core/utils/UpdateChecker.java
Comment (MrNemo64) [reply core/src/main/java/me/lokka30/treasury/plugin/core/utils/UpdateChecker.java:137]: I see that but 
```Java
EconomyProvider provider = providerEconomyProvider.provide();
EconomyAPIVersion providerVersion = provider.getSupportedAPIVersion();
EconomyAPIVersion latestVersion = EconomyAPIVersion.getCurrentAPIVersion();
```
don't seem to modify the parameters nor do anything else than get the parameters for `handleAPIVersioning` so we are callaing the same method with the same values on the `if` and the `else`

File: core/src/main/java/me/lokka30/treasury/plugin/core/utils/UpdateChecker.java
Comment (MrNemo64) [reply core/src/main/java/me/lokka30/treasury/plugin/core/utils/UpdateChecker.java:161]: log levels indicate the severity of the message. They are used to inform not to color code

File: core/src/main/java/me/lokka30/treasury/plugin/core/utils/UpdateChecker.java
Comment (MrIvanPlays) [reply core/src/main/java/me/lokka30/treasury/plugin/core/utils/UpdateChecker.java:137]: this time I am not fully sarcastic but you're blind.

if you know how `if (...) { ... } else { ... }` works (and I can't really see that you know by this comment) you know that only 1 of them would be called at certain state. And I can also see that you haven't checked what the if statement's condition is. 

I am not going to respond to this comment any further, this conversation in my opinion is useless.

File: core/src/main/java/me/lokka30/treasury/plugin/core/utils/UpdateChecker.java
Comment (MrIvanPlays) [reply core/src/main/java/me/lokka30/treasury/plugin/core/utils/UpdateChecker.java:161]: my experiments conclude that server owners/administrators tend to be taken more by the color of the message rather than the log level.

File: core/src/main/java/me/lokka30/treasury/plugin/core/utils/PluginVersion.java
Comment (MrNemo64) [reply core/src/main/java/me/lokka30/treasury/plugin/core/utils/PluginVersion.java:41]: I'm not very familiar with regex but this should do it
```Java
private static final Pattern VERSION_PATTERN = Pattern.compile("\\d+\\.\\d+\\.\\d+-[0-9A-Fa-f]{7}-(SNAPSHOT|RELEASE)");
...
if(!VERSION_PATTERN.matcher(input).matches()) {
  throw new IllegalArgumentException("Your current Treasury version is not formatted correctly! Please inform a Treasury maintainer as soon as possible."):
}
```
The pattern searches this:
`\d+\.`: Any decimal character (1 or more) followed by a '.'
`\d+\.`: Any decimal character (1 or more) followed by a '.'
`\d+-`: Any decimal character (1 or more) followed by a '-'
`[0-9A-Fa-f]{7}-`: Seven hexadecimal characters followed by a '-'
`(SNAPSHOT|RELEASE)`. Literal string "SNAPSHOT" or literal string "RELEASE"

File: core/src/main/java/me/lokka30/treasury/plugin/core/utils/UpdateChecker.java
Comment (MrNemo64) [reply core/src/main/java/me/lokka30/treasury/plugin/core/utils/UpdateChecker.java:137]: If in the if and the else the exact same line of code is going to be executed, regardless of if the `providerEconomyProvider` is null or not, with the same parameters, why not put it outside of the if?

File: core/src/main/java/me/lokka30/treasury/plugin/core/utils/PluginVersion.java
Comment (MrIvanPlays) [reply core/src/main/java/me/lokka30/treasury/plugin/core/utils/PluginVersion.java:41]: The online tester didn't work, however I've tested the following code:

```java
  public static void main(String[] args) {
    Pattern VERSION_PATTERN = Pattern.compile("\\d+\\.\\d+\\.\\d+-[0-9A-Fa-f]{7}-(SNAPSHOT|RELEASE)");
    String matchAgainst = "1.0.0-003b4a0-SNAPSHOT";

    Matcher matcher = VERSION_PATTERN.matcher(matchAgainst);
    if (!matcher.matches()) {
      System.err.println("Nope.");
      return;
    }

    System.out.println("Match successful");
  }
```

And it looks like it works: 
img

I'm implementing it.

**Comment by MrIvanPlays** [comment]
@MrNemo64 if you don't have anything else to say in the next 30 minutes I'm merging.

**Comment by MrNemo64** [comment]
> @MrNemo64 if you don't have anything else to say in the next 30 minutes I'm merging.

Don't think I have anything more to say

**Pull request merged**
